src/thor/costmatrix.cc

```diff
@@ -152,70 +150,87 @@ void CostMatrix::SourceToTarget(Api& request,
-          if (remaining_targets_ > 0) {
-            remaining_targets_--;
+          // in any case make sure this was the last time we looked at this target
+          locs_status_[MATRIX_REV][i].threshold = -1;
+          if (locs_remaining_[MATRIX_REV] > 0) {
+            locs_remaining_[MATRIX_REV]--;
           }
         }
       }
     }
 
-    // Iterate all source locations in a forward search
-    for (uint32_t i = 0; i < source_count_; i++) {
-      if (source_status_[i].threshold > 0) {
-        source_status_[i].threshold--;
-        ForwardSearch(i, n, graphreader, time_infos[i], invariant);
-        if (source_status_[i].threshold == 0) {
-          for (uint32_t target = 0; target < target_count_; target++) {
-            //  Get all sources remaining for the destination
-            auto& sources = target_status_[target].remaining_locations;
+    for (uint32_t i = 0; i < locs_count_[MATRIX_FORW]; i++) {
+      if (locs_status_[MATRIX_FORW][i].threshold > 0) {
+        locs_status_[MATRIX_FORW][i].threshold--;
+        Expand<MatrixExpansionType::forward>(i, n, graphreader, time_infos[i], invariant);
+        // if we exhausted this search
+        if (locs_status_[MATRIX_FORW][i].threshold == 0) {
+          for (uint32_t target = 0; target < locs_count_[MATRIX_REV]; target++) {
+            // if we still didn't find the connection between this pair
+            auto& sources = locs_status_[MATRIX_REV][target].unfound_connections;
             auto it = sources.find(i);
             if (it != sources.end()) {
+              // remove the source so we don't come here again
               sources.erase(it);
-              if (sources.empty() && target_status_[target].threshold > 0) {
-                target_status_[i].threshold = -1;
-                if (remaining_targets_ > 0) {
-                  remaining_targets_--;
+              // if there's no more sources and the current target has not exhausted
+              // we update the target's threshold so that it doesn't enter this outer "if" statement
+              // anymore
+              if (sources.empty() && locs_status_[MATRIX_REV][target].threshold > 0) {
+                // TODO(nils): shouldn't we extend the search here similar to bidir A*
+                //   i.e. if pruning was disabled we extend the search in the other direction
+                locs_status_[MATRIX_REV][i].threshold = -1;
+                if (locs_remaining_[MATRIX_REV] > 0) {
+                  locs_remaining_[MATRIX_REV]--;
                 }
               }
             }
           }
-          source_status_[i].threshold = -1;
-          if (remaining_sources_ > 0) {
-            remaining_sources_--;
+          // in any case make sure this was the last time we looked at this source
+          locs_status_[MATRIX_FORW][i].threshold = -1;
+          if (locs_remaining_[MATRIX_FORW] > 0) {
+            locs_remaining_[MATRIX_FORW]--;
           }
         }
       }
     }
```

Comment:
ok i got to here, am i correct that nothing really changed in either of these loops, was just renaming and comments

Reply:
correct. the whole costmatrix was 95% pure refactor, the remaining 5% the deadend logic, which is also the only thing I wrote additional tests for. I'd like to add some tests comparing the algos too, making sure they report the same duration/distance (+/- some minor threshold)